compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,119 @@
+use rustc_abi::Abi;
+use rustc_middle::ty::{self, Instance, InstanceKind, Ty, TyCtxt};
+use rustc_span::{def_id::DefId, Span, Symbol};
+use rustc_target::abi::call::{FnAbi, PassMode};
+
+use crate::errors::{AbiErrorDisabledVectorTypeCall, AbiErrorDisabledVectorTypeDef};
+
+const SSE_FEATURES: &'static [&'static str] = &["sse", "sse2", "ssse3", "sse3", "sse4.1", "sse4.2"];
+const AVX_FEATURES: &'static [&'static str] = &["avx", "avx2", "f16c", "fma"];
+const AVX512_FEATURES: &'static [&'static str] = &[
+    "avx512f",
+    "avx512bw",
+    "avx512cd",
+    "avx512er",
+    "avx512pf",
+    "avx512vl",
+    "avx512dq",
+    "avx512ifma",
+    "avx512vbmi",
+    "avx512vnni",
+    "avx512bitalg",
+    "avx512vpopcntdq",
+    "avx512bf16",
+    "avx512vbmi2",
+];
+
+fn do_check_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    abi: &FnAbi<'tcx, Ty<'tcx>>,
+    target_feature_def: DefId,
+    emit_err: impl Fn(&'static str),
+) {
+    // This check is a no-op on non-x86, at least for now.
+    if tcx.sess.target.arch != "x86" && tcx.sess.target.arch != "x86_64" {
+        return;
+    }
+    let codegen_attrs = tcx.codegen_fn_attrs(target_feature_def);
+    for arg_abi in abi.args.iter().chain(std::iter::once(&abi.ret)) {
+        let size = arg_abi.layout.size;
+        if matches!(arg_abi.layout.abi, Abi::Vector { .. })
+            && matches!(arg_abi.mode, PassMode::Direct(_))
+        {
+            let features: &[_] = match size.bits() {
+                x if x <= 128 => &[SSE_FEATURES, AVX_FEATURES, AVX512_FEATURES],
+                x if x <= 256 => &[AVX_FEATURES, AVX512_FEATURES],
+                x if x <= 512 => &[AVX512_FEATURES],
+                _ => {
+                    panic!("Unknown vector size for x86: {}; arg = {:?}", size.bits(), arg_abi)
+                }
+            };
+            let required_feature = features.iter().map(|x| x.iter()).flatten().next().unwrap();
+            if !features.iter().map(|x| x.iter()).flatten().any(|feature| {
+                let required_feature_sym = Symbol::intern(feature);
+                tcx.sess.unstable_target_features.contains(&required_feature_sym)
+                    || codegen_attrs.target_features.contains(&required_feature_sym)
+            }) {
+                emit_err(required_feature);
+            }
+        }
+    }
+}
+
+/// Checks that the ABI of a given instance of a function does not contain vector-passed arguments
+/// or return values for which the corresponding target feature is not enabled.
+pub fn check_instance_abi<'tcx>(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) {
+    let InstanceKind::Item(item_def) = instance.def else {
+        return;
+    };
+
+    let param_env = tcx.param_env(item_def);
```

Comment:
During monomorphization you can always use `ParamEnv::reveal_all`.

(This here seems like a likely candidate for the source of the ICE. The oaram_env you are asking for is *inside* the item but the item you got is already monomorphized.)